### PR TITLE
feat: remove session id from session data

### LINKED ISSUE #52: Exposing encryptedSessionId leaks secret to database
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.25.3

### Plugin version

6.3.0

### Node.js version

14.15.1

### Operating system

Windows

### Operating system version (i.e. 20.04, 11.3, 10)

10

### Description

Compared to express session, fastify session exposes the encrypted session id to the database. This could be used to determine the session secret given a database leak and thus leaking temporary password equivalent. It also wastes space by including the sessionId twice and expiry date three times.

express-session
```json
"_id":"abcdefghijklmnopqrstuvwxyz",
"expires":"2022-03-08T09:43:51.997+00:00",
"session":{
    "cookie":{
        "originalMaxAge":"5184000000",
        "expires":"2022-03-08T09:43:35.786+00:00",
        "secure":false,
        "httpOnly":true,
        "domain":null,
        "path":"/",
        "sameSite":null
    }
}
```
fastify/session
```json
"_id":"abcdefghijklmnopqrstuvwxyz",
"expires":"2022-03-08T09:43:51.997+00:00",
"session":{
    "expires":"2022-03-08T09:43:51.997+00:00",
    "cookie":{
        "originalMaxAge":"5184000000",
        "expires":"2022-03-08T09:43:35.786+00:00",
        "secure":false,
        "httpOnly":true,
        "domain":null,
        "path":"/",
        "sameSite":"strict"
    },
    "sessionId":"abcdefghijklmnopqrstuvwxyz",
    "encryptedSessionId":"qwertyuiopasdfghjklzxcvbnm"
}
```

### Steps to Reproduce

```js
import fp from 'fastify-plugin';
import cookie from 'fastify-cookie';
import session from '@fastify/session';

// https://github.com/jdesboeufs/connect-mongo/issues/441
import _ from 'express-session';
import MongoStore from 'connect-mongo';

export default fp(async (fastify, options) => {
    if (!fastify.dbClient) {
        throw new Error('Must register db plugin');
    }

    fastify.register(cookie);
    fastify.register(session, {
        secret: process.env.SECRET,
        saveUninitialized: false,
        cookieName: '__Host-sessionId',
        cookie: {
            sameSite: 'strict',
            httpOnly: true,
            secure: false
        },
        store: MongoStore.create({
            stringify: false,
            client: /* ... */
        })
    });
});
```

### Expected Behavior

Do not write encryptedSessionId to database.

### PR DESCRIPTION
**Pull request by SimenB**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Fixes #52
Closes #70

I'm not 100% sure about the test changes, would like a second pair of eyes

### CODE REVIEW

File: test/base.test.js [hunk 1]
```diff
@@ -50,8 +50,10 @@ test('should support multiple secrets', async (t) => {
 
   const plugin = fastifyPlugin(async (fastify, opts) => {
     fastify.addHook('onRequest', (request, reply, done) => {
+      const expires = Date.now() - 1000
       request.sessionStore.set('aYb4uTIhdBXCfk_ylik4QN6-u26K0u0e', {
-        expires: Date.now() + 1000
```
Comment (SimenB) [inline test/base.test.js:54]: I don't understand this test, so this change is probably wrong

File: test/base.test.js [hunk 2]
```diff
@@ -99,7 +101,6 @@ test('should set session cookie using the default cookie name', async (t) => {
     fastify.addHook('onRequest', (request, reply, done) => {
       request.sessionStore.set('Qk_XT2K7-clT-x1tVvoY6tIQ83iP72KN', {
         expires: Date.now() + 1000,
-        sessionId: 'Qk_XT2K7-clT-x1tVvoY6tIQ83iP72KN',
```
Comment (SimenB) [inline test/base.test.js:102]: there's only a getter, no setter. Probably makes this a breaking change even though _changing_ the session id has never been supported I think?

It's possibly breaking regardless as somebody have a custom store that relies on this data being present? not sure

File: lib/session.js [hunk 3]
```diff
@@ -29,10 +30,8 @@ module.exports = class Session {
     this[addDataToSession](prevSession)
     this[requestKey] = request
     this.touch()
-    if (!this.sessionId) {
```
[the remark this answers is not in the document]
Comment (SimenB) [reply lib/session.js:32]: I just understood it was previously set by the `this[addDataToSession]`, so the `if` made sense. That's no longer done though, so removing the `if` is correct

File: test/base.test.js (hunk 2 above)
Comment (SimenB) [reply test/base.test.js]: I could revert this after updating the allowlist in `this[addDataToSession]`. It's now ignored instead

File: lib/session.js [hunk 4]
```diff
@@ -172,10 +167,18 @@ module.exports = class Session {
...
+  }
+
+  get encryptedSessionId () {
+    return this[encryptedSessionIdKey]
+  }
+
+  static restore (request, idGenerator, cookieOpts, secret, prevSession, sessionId) {
+    const restoredSession = new Session(request, idGenerator, cookieOpts, secret, prevSession, sessionId)
     const restoredCookie = new Cookie(cookieOpts)
-    restoredCookie.expires = new Date(prevSession.cookie.expires)
```
[the remark this answers is not in the document]
Comment (mcollina) [reply lib/session.js:178]: what should be done about this?

[the remark this answers is not in the document]
Comment (SimenB) [reply lib/session.js:178]: either works I think, but I think it's weird to look at the expiry of the cookie and not the expiry itself.

That said, `express-session` _only_ looks at the cookie data, and have not hoisted `expires` up into its own top level thing. So we could do the same here and just get rid of the top level `expires`

[the remark this answers is not in the document]
Comment (mcollina) [reply lib/session.js:178]: Go for that!

File: test/session.test.js [hunk 5]
```diff
@@ -36,9 +36,11 @@ test('should destroy the session', async (t) => {
 })
 
 test('should add session.encryptedSessionId object to request', async (t) => {
-  t.plan(2)
+  t.plan(3)
   const port = await testServer((request, reply) => {
     t.truthy(request.session.encryptedSessionId)
+    // serialize, then deserialize to make sure it's gone
+    t.falsy(JSON.parse(JSON.stringify(request.session)).encryptedSessionId)
```
[the remark this answers is not in the document]
Comment (mcollina) [reply test/session.test.js:43]: Seems good to me!

File: test/base.test.js (hunk 1 above)
Comment (SimenB) [reply test/base.test.js:54]: I need to dig into the test to understand what it's supposed to test.

**Comment by sarneeh** [comment]
Hey! Can I help somehow to finish this PR? I'd need this change as well 🙏

**Comment by SimenB** [comment]
@sarneeh sure, go for it!

We tried rolling this out at work, and had to revert due to some sessions _seemingly_ not updating correctly. Unfortunately I haven't found the time yet to dig into it, which also means this PR is on the backburner for the moment (and now I'm on vacation, so won't get to it for at least a few more weeks). However, feel free to pick this up in the meantime!

**Review by rclmenezes — COMMENTED** [review]
Rebased + fixed some issues in: https://github.com/fastify/session/pull/114

**Comment by Uzlopak** [comment]
closing this in favor of #114

**Pull request closed without merging**